### DISCUSSION TITLE
Devices with static WAN IP address needn't ping to a Super Node

### DISCUSSION THREAD

**Discussion by photosssa** [buckyos/CYFS#76] [Ideas]
# If a device connected to ethernet with NAT,  to establish a P2P connection to this device from any device, Super Node(SN) is needed;
+ Device got its WAN IP address with NAT from ping/pong with a SN;
+ Device keep its WAN IP address alive with ping/pong with a SN continuously;
+ Device got a handshake transited from SN for NAT may drop packet from IP addresses without a former deliver record;

# If a device has static WAN IP address,  means:
+ Device administrator knows its WAN IP address;
+ Device WAN IP address keeps unchanged(static);
+ Packets from any location sent to this WAN IP address reaches;
SN is no more needed to establish a connection to devices with static WAN address;

# How to config WAN IP address and no pinging SN in CYFS
## Config endpoints for CYFS Stack
1. Device has WAN IP address (without NAT): Config this WAN address for CYFS stack, socket bound to this WAN address; 
2. Device has a mapped static WAN IP address(with NAT): 
+ Config both local address and its mapped WAN address for CYFS stack, socket bound to local address
+ Config local address and a its mapped port for CYFS stack, socket bound to local address, then ping/pong with SN once, got its mapped WAN address combining WAN IP and mapped port;
    
## Config optional super nodes for CYFS stack
+ Config optional super nodes for CYFS stack, select one or more SN from those to ping/pong with; 
+ If device needn't SN, config an empty optional super nodes; but CYFS stack need other ways to announce its endpoints(i.e. static configed known device; object Key-value storage on block chain);
 
## Config known super nodes for CYFS stack
1. To establish a connection from device with static WAN address to other devices, target device's pinging SN is still needed;
2. Config known super nodes for CYFS stack to find target device's ping SN in those; it differs from optional super nodes;

**Comment by lurenpluto** [comment]
Support for NO-SN mode has been added into cyfs-stack, see the issue below
https://github.com/buckyos/CYFS/issues/56

**Comment by streetycat** [comment]
So, does that means the `sn` is not needed in `Pure LAN` ?

**Reply by lurenpluto** [reply]
In a way, it can be said that

At present, SN is not only used for `connection service` of all peers (devices, mainly devices located in LAN) in cyfs system, but also plays a role of `service discovery` mechanism of devices, peers upload devices when pinging, other peers can query back by selecting the correct sn when connecting. device, if the device itself is configured with WAN address, then you can directly use this (device.desc+wan addr) to initiate the connection

**Reply by streetycat** [reply]
Does it mean `SN` can also help the peers to publish the changed `IP`, and others can connect it with the latest `IP`.

And the `wan addr` is only a flag to tell others that the address can be connected directly.

**Reply by lurenpluto** [reply]
In cyfs network, bdt protocol connection is based on Device object (representing a peer), device can configure peer's ip address list, if this contains the type of WAN ip address, then bdt will try to use this WAN ip to connect directly when connecting to this device in preference

SN also plays the role of service discovery in cyfs network to some extent, and can be used to report and query Device Object based on ping

**Reply by streetycat** [reply]
I see, it's interesting
